### ISSUE TITLE
Game.agf file randomly reorders lines

### ISSUE THREAD

**Issue by messengerbag** [adventuregamestudio/ags#1282]
**Describe the bug**
When saving the game project, the Game.agf file will sometimes change, apparently at random, by reordering some lines/tags. For example:

```
    <AudioClips>
        <!-- Bunch of nested tags -->
                  <AudioClip>
                    <SourceFileName xml:space="preserve">D:\Music\AGS\agsawards2015\x. Acceptance Speech No Space Quest.mp3</SourceFileName>
                    <ScriptName>aIntro</ScriptName>
```
Randomly changed to:

```
    <AudioClips>
        <!-- Bunch of nested tags -->
                  <AudioClip>
                    <ScriptName>aIntro</ScriptName>
                    <SourceFileName xml:space="preserve">D:\Music\AGS\agsawards2015\x. Acceptance Speech No Space Quest.mp3</SourceFileName>
```
(As you can see, the only change is that the `<SourceFileName />` and `<ScriptName />` tags change place.)

This adds junk to the version history and makes it harder to diff versions of the file in source control. 

**AGS Version**
Issue observed with 3.4.3

**To Reproduce**
1. Keep an AGS project in source control (git), making sure that Game.agf is treated as text
2. Save the game project and commit it.
3. Save the project again, check for changes.

**Expected behavior**
The Game.agf file should always be saved with tags in the same order, so that the file only changes when there is an actual change in content.

**Comment by ericoporto** [comment]
Not completely sure but maybe it's just sorting the `properties` after the line here https://github.com/adventuregamestudio/ags/blob/638b50fa29f15af13297edd708afedc59d56db07/Editor/AGS.Types/SerializeUtils.cs#L86

**Comment by ivan-mogilko** [comment]
In my memory this happened only to AudioTypes or AudioClips, which is rather curious. Maybe there's something in its properties or the way these particular types are written down.

**Comment by persn** [comment]
I seem to have reproduced it somehow. I'll do some debugging

**Comment by persn** [comment]
The documentation for `GetProperties()` says it all

https://docs.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=net-5.0
> The GetProperties method does not return properties in a particular order, such as alphabetical or declaration order. Your code must not depend on the order in which properties are returned, because that order varies.

Like @ericoporto says, we need to add our own order enforcing. I imagine that sorting it alphabetically or whatever will cause volatile changes in the `Game.agf` for people using version control. Maybe fix this in `ags4`?

**Comment by ericoporto** [comment]
Hey, I think it should be fine to break a little the version control if done on current master (3.6.0.X). @persn when I answered I was just browsing the code on my phone, did it really worked? :D

**Comment by ivan-mogilko** [comment]
Well, this is not a breaking change per se, it does not change data format, only reorders fields in the format that does not depend on fields order. I guess it's okay for 3.6.0.

**Comment by messengerbag** [comment]
> I imagine that sorting it alphabetically or whatever will cause volatile changes in the Game.agf for people using version control.

I might be misunderstanding you, but it seems to me that this would _not_ cause volatile changes because the tag names in the agf file are stable, so sorting alphabetically will keep the same order each time. (Though if someone upgrades the project from an older version then it might introduce a lot of changes the first time the game is saved under the new version; this does not seem like a big problem to me: a single commit with a bunch of predictable reordering in Game.agf is better than random reordering happening on almost _every_ commit.)

In other words, my understanding is that the proposed solution should work fine. (It would need to be tested, of course.)

**Comment by persn** [comment]
Someone who attempts upgrading with work in progress will have a nasty merge conflict I gotta say, they can probably blame themselves for upgrading with WIP though.

![image](https://user-images.githubusercontent.com/1510803/119193959-ab74b780-ba82-11eb-8cfc-3c555bad1f93.png)

**Comment by messengerbag** [comment]
This is only a conflict if two people are editing the project in parallel; otherwise it's just a change. And it can easily be resolved by having both of them update to the new version, which will also prevent any such conflicts going forward.

**Comment by ivan-mogilko** [comment]
@persn why is the merge conflict so huge though? Should not it only include inner fields in the nodes? Or node order also changes?

**Comment by messengerbag** [comment]
Looking at the screenshot, it looks odd that all the fields in `<AudioClip>` are listed alphabetically except for `<FileLastModifiedDate>` which comes at the end. Is that not one of the properties returned by `obj.GetType().GetProperties()`?

**Comment by persn** [comment]
@ivan-mogilko What is upper nodes here? Is it AudioClips? SubFolders? At any rate this is pretty much what I expected to happen and is why I referred to the change as volatile, people in version control are going to have a surprise when they check the diff after upgrade, and potentially nasty merge conflicts that are impossible to resolve if they are careless. The fix is applied to the root of the xml serializer mechanism so it makes perfectly sense that it affects the entire file.

@messengerbag Not every property is serialized through reflection unfortunately, There are lots of cases in the source code where certain properties have manual serialization for whatever reason. If we check `<AudioClip><FileLastModifiedDate>` It has custom serilizalization to specify [date format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings#UniversalSortable), and thus it gets appended to the end of every other property.

That's fine though, isn't it? The goal is to achieve a constant order of appearance, not alphabetical ordering.

```csharp
public void ToXml(XmlTextWriter writer)
{
    SerializeUtils.SerializeToXML(this, writer, false); // Alphabetical ordering happens here
    writer.WriteElementString("FileLastModifiedDate", _fileLastModifiedDate.ToString("u"));
    writer.WriteEndElement();
}
```

**Comment by persn** [comment]
If absolute alphabetical ordering is what we want for whatever reason we could make a override class for `XmlTextWriter` and execute `WriteElementString` lazily and ordered. That way we can ensure it always happen when xml is written

**Comment by messengerbag** [comment]
> That's fine though, isn't it? The goal is to achieve a constant order of appearance, not alphabetical ordering.

Yes, I just wanted to ensure that it was actually working correctly. As long as the order is deterministic and stable, it should be fine.

> What is upper nodes here? Is it AudioClips? SubFolders? At any rate this is pretty much what I expected to happen

I've never noticed nodes at this level of the hierarchy moving around, and since the order they typically appear in in the file more or less matches the order in the project explorer, I had thought that they were already in some fixed order. If this is not the case, then you're right; sorting them is called for. (And even if they are already in some fixed order, we might conclude that reordering them is better because this approach benefits from simplicity.)

> people in version control are going to have a surprise when they check the diff after upgrade, and potentially nasty merge conflicts that are impossible to resolve if they are careless.

Is it impossible to resolve? If they get into a merge conflict, can't they just both upgrade their project to the new version, and that way they'll both be consistent?

**Comment by persn** [comment]
AudioClips and other entities will appear in order you see in solution explorer I think, elements within each entity will be ordered alphabetically first, then custom handled elements are placed at the bottom.

> Is it impossible to resolve? If they get into a merge conflict, can't they just both upgrade their project to the new version, and that way they'll both be consistent?

Perhaps, good luck to anyone why tries it.

I see now that the aggresiveness of the diff output depends the diffing algorithm, opening up the same in Git Extensions gives nicer output.

![image](https://user-images.githubusercontent.com/1510803/119234606-21cef380-bb2f-11eb-94b0-2e56a63b4792.png)

**Comment by ivan-mogilko** [comment]
Yes, the last screenshot looks more like it should. I was mostly surprised why the parent nodes were detected as changed in the previous case, even though they seemed unchanged.
